java/java.lsp.server/vscode/src/explorer.ts

```diff
@@ -154,42 +169,131 @@ class VisualizerProvider extends vscode.Disposable implements CustomizableTreeDa
+   */
+  private releaseVisualizersAndFire(list : Visualizer[] | undefined) {
+    if (!list) {
+      list = Array.from(this.delayedFire);
+    }
+    if (doLog) {
+      this.log.appendLine(`Done with ${this.visualizerList(list)}`);
+    }
+    // v can be in list several times, each push increased its counter, so we need to decrease it.
+    for (let v of list) {
+      if (this.treeData?.get(Number(v.id || -1)) === v) {
+        if (--v.pendingQueries) {
+          if (doLog) {
+            this.log.appendLine(`${v.idstring()} has pending ${v.pendingQueries} queries`);
+          }
+          continue;
+        }
+        if (v.pendingChange) {
+          if (doLog) {
+            this.log.appendLine(`Fire delayed change on ${v.idstring()}`);
+          }
+          this.fireItemChange(v);
+          v.pendingChange = false;
+        }
+      }
+      this.delayedFire.delete(v);
     }
+    if (doLog) {
+      this.log.appendLine("Pending queue: " + this.visualizerList(Array.from(this.delayedFire)));
+      this.log.appendLine("---------------");
+    }
+  }
+
+  /**
+   * Should wrap calls to NBLS for individual visualizers (info, children). Puts visualizer on the delayed fire list.
+   * Must be itself wrapped in wrap() -- wrap(... queryVisualizer()).
+   * @param element visualizer to be queried, possibly undefined (new item is expected)
+   * @param fn code to execute
+   * @returns code's result
+   */
+  async queryVisualizer<X>(element : Visualizer | undefined, pending : Visualizer[], fn : () => Promise<X>) : Promise<X> {
+    if (!element) {
+      return fn();
+    }
+    this.delayedFire.add(element);
+    pending.push(element);
+    element.pendingQueries++;
+    if (doLog) {
+      this.log.appendLine(`Delaying visualizer ${element.idstring()}, queries = ${element.pendingQueries}`)
+    }
+    return fn();
+  }
+
+  async getTreeItem2(element: Visualizer): Promise<vscode.TreeItem> {
+    const n = Number(element.id);
     if (this.decorators.length == 0) {
-      return element;
+     return element;
     }
     let list : TreeItemDecorator<Visualizer>[] = [...this.decorators];
     
-    function f(item : vscode.TreeItem) : vscode.TreeItem | Thenable<vscode.TreeItem> {
+    async function f(item : vscode.TreeItem) : Promise<vscode.TreeItem> {
       const deco = list.shift();
       if (!deco) {
-        return item;
+       return item;
       }
       const decorated = deco.decorateTreeItem(element, item);
       if (decorated instanceof vscode.TreeItem) {
           return f(decorated);
       } else {
-          return (decorated as Thenable<vscode.TreeItem>).then(f);
+         return (decorated as Thenable<vscode.TreeItem>).then(f);
       }
     }
-
     return f(element.copy());
   }
 
+  delayedFire : Set<Visualizer> = new Set<Visualizer>();
+
   async fetchItem(n : number) : Promise<Visualizer> {
+    let refreshVis = this.treeData.get(n);
```

Comment:
`refreshVis` is not used anywhere.

Reply:
If refreshVis exists, it's not necessary to create it below, is that right?

Reply:
Right it is a leftover. Updates of the existing `Visualizer` instances are done in `fetchItem` callers.

Reply:
Thanks.